### DISCUSSION TITLE
How to change layer mode after request

### DISCUSSION THREAD

**Discussion by marc-mabe** [unpoly/unpoly#491] [Q&A] [answered]
If a session has been timed out the server responses with 401 Unauthorized and a user friendly message.
Now if this happens on opening a layer I would like to abort the layer and instead open a modal using the content of the response matching target.

Right now I have this:
```js
up.on('up:request:loaded', function(event: any) {
    if (event.response.status === 401) {
        up.fragment.abort(event.target);
        up.layer.open({
            content: event.response.text,
            mode: 'modal',
            target: 'main',
            dismissable: false,
        });
    }
});
```

and the response looks like this:
```html
<html>
    <head>...</head>
    <body>
        <nav>...</nav>
        <main>...</main>
    </body>
</html>
```

My problem is now that the full reponse html is used and put into the main target of the current page but I want to replace the current main target with the main element of the response.

**Comment by marc-mabe** [comment]
I'm doing it this way now - hope it makes sense:
```js
up.on('up:request:loaded', function(event: any) {
    if (event.response.status === 401) {
        try {
            const html = event.response.text;
            const htmlDoc = new DOMParser().parseFromString(html, 'text/html');

            up.fragment.abort(event.target);
            up.layer.open({
                document: htmlDoc,
                mode: 'modal',
                target: 'main',
                dismissable: false,
            });
        } catch (e) {
            console.error('Failed parsing 401 response as HTML');
        }
    }
});
```

But still wondering why the `target` is ignored on passing `content`.

**Comment by adam12** [comment, marked as answer]
Maybe look to see if [`up:fragment:loaded`](https://unpoly.com/up:fragment:loaded) works better for you? Then you can manipulate the [`renderOptions`](https://unpoly.com/up:fragment:loaded#example-changing-render-options).

**Comment by marc-mabe** [comment]
@adam12 Thank your your helpful comment. This is actually a really nice solution :+1:  just that it does not work in my (deeper) case as the login needs to redirect the user to another service which does not work in an async modal. So I have to abort the current, create a new modal and let the user be redirected.

Am I right that `up:fragment:loaded` does not catch all cases a request could be made to the server which could respond with a 401?

Anyway as your documentation links are matching my original question I'm accepting your answer :100:

**Reply by adam12** [reply]
> Am I right that `up:fragment:loaded` does not catch all cases a request could be made to the server which could respond with a 401?

Unfortunately, I cannot answer that with any certainty, but I suspect it's true. I generally find 401 statuses to be very special cases because of how the browser built-ins work. Perhaps Henning knows more as he's familiar with that part of the code.
